test/Cardano/FeeSpec.hs

```diff
@@ -460,6 +466,54 @@ propDistributeFeeIsFair (fee, coins) = (.&&.)
         * fromIntegral (getFee fee)
         % fromIntegral (sum $ getCoin <$> coins)
 
+-- | Verify that fees are distributed optimally across coins, such that the
+-- absolute deviation from the ideal (unrounded) fee distribution is minimal.
+propDistributeFeeOptimal
+    :: Fee
+    -> NonEmpty Coin
+    -> Property
+propDistributeFeeOptimal fee coins = property $
+    computeDeviation (fst <$> distributeFee fee coins)
+        `shouldBe` minimumPossibleDeviation
+  where
+    -- Compute the deviation of a fee portion distribution from the ideal
+    -- unrounded fee portion distribution.
+    computeDeviation :: NonEmpty Fee -> Rational
+    computeDeviation feesRounded = F.sum $
+        NE.zipWith (\(Fee f) u -> abs (fromIntegral f - u))
+            feesRounded
+            idealUnroundedDistribution
+
+    -- The minimum deviation across all possible distributions for the given
+    -- fee and set of coins.
+    minimumPossibleDeviation :: Rational
+    minimumPossibleDeviation =
+        F.minimum $ computeDeviation <$> allPossibleDistributions
+
+    -- The set of all possible fee distributions for the given fee and coins.
+    allPossibleDistributions :: [NonEmpty Fee]
+    allPossibleDistributions = filter isValidDistribution $ NE.zipWith
+        (\f roundDir -> Fee $ fromIntegral @Integer $ round roundDir f)
+        idealUnroundedDistribution <$> allPossibleRoundings
+      where
+        -- Indicates whether the given distribution has the correct total fee.
+        isValidDistribution :: NonEmpty Fee -> Bool
+        isValidDistribution r = sum (getFee <$> r) == getFee fee
+
+        -- All possible ways to round an unrounded fee distribution.
+        allPossibleRoundings :: [NonEmpty RoundingDirection]
+        allPossibleRoundings = traverse (const [RoundUp, RoundDown]) coins
+
+    -- The ideal unrounded fee distribution.
```

Comment:
so ideal fee is when we treat (ie., sum) all coins as one coin and calculate fee on it. not afraid here we will exceed 45b ada and it will affect somehow this?

Reply:
> not afraid here we will exceed 45b ada and it will affect somehow this?

I think this shouldn't be a problem, as:

1. all internal calculations use either `Integer` or `Rational`, only converting back to `Fee` at the end.
2. all the `Fee` portions should be less than or equal to the initial `Fee` value.